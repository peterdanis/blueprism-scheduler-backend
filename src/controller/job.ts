import Job from "../entity/Job";
import log from "../utils/logger";
import Schedule from "../entity/Schedule";
import ScheduleTask from "../entity/ScheduleTask";
import { v4 as uuid } from "uuid";

let checking = false;
let firstRun = true;
const runningJobs: Job[] = [];

/**
 * Returns array of unique resource ids found in given job array
 */
const getResourcesFromJobs = (jobs: Job[]): number[] => {
  const reducer = (
    jobArr: number[],
    { runtimeResource: { id } }: Job,
  ): number[] => {
    if (jobArr.includes(id)) {
      return jobArr;
    }
    return [id, ...jobArr];
  };
  return jobs.reduce(reducer, []);
};

/**
 * Sorts jobs by priority, from lowest to highest. Jobs with lower priority should go first.
 */
const orderByPriority = (jobs: Job[]): Job[] => {
  return jobs.sort((first, second) => first.priority - second.priority);
};

const filterByResource = (id: number, jobs: Job[]): Job[] => {
  return jobs.filter(({ runtimeResource }) => runtimeResource.id === id);
};

/**
 * Returns true if no jobs are in "running" status for given resource id.
 */
const isResourceFree = (id: number, jobs: Job[]): boolean => {
  return (
    filterByResource(id, jobs).filter(({ status }) => status === "running")
      .length === 0
  );
};

/**
 * Adds job to queue and returns the reference to it.
 */
export const addJob = async (schedule: Schedule): Promise<Job> => {
  const { id, rule, priority, runtimeResource } = schedule;
  log(`Adding job with id ${id} and rule ${rule} to jobs`);

  const job = Job.create({
    addTime: new Date(),
    priority,
    runtimeResource,
    schedule,
    status: "waiting",
  });
  return job.save();
};

const run = async (job: Job): Promise<void> => {
  try {
    const {
      sessionId,
      step,
      subStep,
      runtimeResource: { id },
      schedule: { scheduleTask },
    } = job;
    // Step is indexed from 1, need to substract 1 to get the correct array index
    if (scheduleTask.length === 0) {
      return;
    }
    const stepDetails = scheduleTask.sort(
      (first, second) => first.step - second.step,
    )[step - 1];
    const {
      delayAfter,
      task: { process, inputs, softTimeout, hardTimeout, name },
    } = stepDetails as ScheduleTask;
    log(`Send request to resource ${id}`);
    throw new Error("test");
  } catch (error) {
    log(error);
    job.message = error.message;
    job.status = "failed";
    job.endTime = new Date();
    await job.save();
  }
};

export const getJobs = async (
  condition?: Partial<Job> | Partial<Job>[],
): Promise<Job[]> => {
  if (condition) {
    return Job.find({ where: condition });
  }
  return Job.find();
};

/**
 * Starts queued jobs on free resources. In case of first run, it will try to restart previously running jobs (e.g. after app crash).
 */
export const startIfAvailable = async (): Promise<void> => {
  if (checking) {
    // Return immediately if the function is already running. This is to prevent accidentally starting multiple jobs on one resource.
    return;
  }
  try {
    checking = true;
    const jobs = await getJobs(
      firstRun ? { status: "running" } : { status: "waiting" },
    );
    log(jobs);
    if (jobs.length === 0) {
      return;
    }
    const uniqueResources = getResourcesFromJobs(jobs);
    // eslint-disable-next-line no-restricted-syntax
    for (const id of uniqueResources) {
      // uniqueResources.forEach((id) => {
      if (firstRun || isResourceFree(id, jobs)) {
        const filteredJobs = filterByResource(id, jobs);
        const job = orderByPriority(filteredJobs)[0];
        if (job) {
          job.status = "running";
          // eslint-disable-next-line no-await-in-loop
          await job.save();
          // Do not await the result
          run(job);
        }
      }
    }
  } catch (error) {
    //
  } finally {
    firstRun = false;
    checking = false;
  }
};

// const softStopJob = () => {};

// const hardStopJob = () => {};
