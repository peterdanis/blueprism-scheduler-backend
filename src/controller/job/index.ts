import { JobRef, run } from "./run";
import Job from "../../entity/Job";
import log from "../../utils/logger";
import Schedule from "../../entity/Schedule";

let checking = false;
let firstRun = true;
let runningJobRef: JobRef[] = [];

/**
 * Returns array of unique resource ids found in given job array
 */
const getResourcesFromJobs = (jobs: Job[]): number[] => {
  const reducer = (acc: number[], val: Job): number[] => {
    const { runtimeResource } = val;
    return acc.includes(runtimeResource.id)
      ? acc
      : [runtimeResource.id, ...acc];
  };
  return jobs.reduce(reducer, []);
};

/**
 * Returns jobs sorted by priority, from lowest to highest. Jobs with lower priority should go first.
 */
const orderByPriority = (jobs: Job[]): Job[] => {
  return jobs.sort((first, second) => first.priority - second.priority);
};

/**
 * Returns jobs for given resource id.
 */
const filterByResource = (id: number, jobs: Job[]): Job[] => {
  return jobs.filter(({ runtimeResource }) => runtimeResource.id === id);
};

/**
 * Returns jobs with given status.
 */
const filterByStatus = (jobStatus: Job["status"], jobs: Job[]): Job[] => {
  return jobs.filter(({ status }) => status === jobStatus);
};

/**
 * Returns true if no jobs are in "running" status for given resource id.
 */
const isResourceFree = (id: number, jobs: Job[]): boolean => {
  return (
    filterByResource(id, jobs).filter(({ status }) => status === "running")
      .length === 0
  );
};

/**
 * Adds job to queue and returns the reference to it.
 */
export const addJob = async (
  schedule: Schedule,
  startReason: string,
): Promise<Job> => {
  const { id, rule, priority, runtimeResource } = schedule;

  const job = Job.create({
    addTime: new Date(),
    priority,
    runtimeResource,
    schedule,
    startReason,
    status: "waiting",
  });
  log.info("Adding schedule jobs", {
    rule,
    scheduleId: id,
  });
  return job.save();
};

export const getJobs = async (
  condition?: Partial<Job> | Partial<Job>[],
): Promise<Job[]> => {
  if (condition) {
    return Job.find({ where: condition });
  }
  return Job.find();
};

/**
 * Starts queued jobs on free resources. In case of first run, it will try to restart previously running jobs (e.g. after app crash).
 */
export const startIfAvailable = async (): Promise<void> => {
  if (checking) {
    // Return immediately if the function is already running. This is to prevent accidentally starting multiple jobs on one resource.
    return;
  }
  try {
    checking = true;
    // Check jobRef array for items marked for deletion
    runningJobRef = runningJobRef.filter((jobRef) => !jobRef.delete);

    const jobs = await getJobs([{ status: "running" }, { status: "waiting" }]);
    if (jobs.length === 0) {
      return;
    }
    const uniqueResources = getResourcesFromJobs(jobs);
    // eslint-disable-next-line no-restricted-syntax
    for (const id of uniqueResources) {
      if (firstRun || isResourceFree(id, jobs)) {
        const thisResourceJobs = filterByResource(id, jobs);
        const runningJobs = filterByStatus("running", thisResourceJobs);
        const waitingJobs = filterByStatus("waiting", thisResourceJobs);
        let job: Job | undefined;
        if (runningJobs.length === 0) {
          [job] = orderByPriority(waitingJobs);
        } else {
          [job] = orderByPriority(runningJobs);
        }
        if (job) {
          job.status = "running";
          job.startTime = new Date();
          // eslint-disable-next-line no-await-in-loop
          await job.save();
          // Do not await the result
          const jobRef = run(job);
          runningJobRef.push(jobRef);
        }
      }
    }
  } catch (error) {
    //
  } finally {
    firstRun = false;
    checking = false;
  }
};
