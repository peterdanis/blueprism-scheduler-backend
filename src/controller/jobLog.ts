import Job from "../entity/Job";
import JobLog from "../entity/JobLog";
import log from "../utils/logger";
import retry from "../utils/retry";

export const transferJob = async (
  job: Job,
  message: string,
): Promise<JobLog> => {
  const { id, schedule, runtimeResource } = job;
  const jobLog = JobLog.create({
    ...job,
    // id: undefined,
    jobId: id,
    message,
    runtimeResourceId: runtimeResource.id,
    scheduleId: schedule.id,
  });
  await retry(() => jobLog.save());
  log.info("Job transferred", { jobId: id, jobLogId: jobLog.id });
  await retry(() => job.remove());
  return jobLog;
};

export const getJobLogs = async (): Promise<void> => {
  //
};
