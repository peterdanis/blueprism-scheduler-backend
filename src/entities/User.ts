import { Column, Entity } from "typeorm";
import Base from "./Base";

@Entity()
export default class User extends Base {
  @Column({ unique: true })
  name!: string;

  @Column({ default: true })
  readonly!: boolean;
}
