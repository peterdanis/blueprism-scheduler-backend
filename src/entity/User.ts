import { Column, Entity } from "typeorm";
import Base from "./Base";

@Entity()
export default class User extends Base {
  @Column({ default: false })
  admin!: boolean;

  @Column({ unique: true })
  name!: string;

  @Column()
  password!: string;
}
