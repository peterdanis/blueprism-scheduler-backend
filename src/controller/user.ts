import { compare, hash } from "bcrypt";
import User from "../entity/User";

const saltRounds = 10;

export const addUser = async (
  name: string,
  password: string,
): Promise<User> => {
  const user = User.create({ name });
  user.password = await hash(password, saltRounds);
  return user.save();
};

export const updateUser = async (): Promise<void> => {
  //
};

export const verifyPassword = async (password: string): Promise<void> => {
  //
};
