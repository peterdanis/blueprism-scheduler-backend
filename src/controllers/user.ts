import { compare, hash } from "bcrypt";
import { createHash } from "crypto";
import User from "../entities/User";
import { v4 as uuid } from "uuid";

const saltRounds = 10;

let userCache: User[] | undefined;

const generateHash = (string: string): string =>
  createHash("sha256").update(string).digest("base64");

export const addUser = async (
  name: string,
  password: string,
): Promise<User> => {
  userCache = undefined;
  const user = User.create({ name });
  user.password = await hash(password, saltRounds);
  await user.save();
  return user;
};

export const getUsers = async (): Promise<User[]> => {
  if (!userCache) {
    userCache = await User.find();
  }
  return userCache;
};

export const getUser = async (
  nameOrId: string | number,
): Promise<User | undefined> => {
  const users = await getUsers();
  const [user] = users.filter((_user) => {
    if (typeof nameOrId === "string") {
      return _user.name === nameOrId;
    }
    return _user.id === nameOrId;
  });
  return user;
};

export const updateUser = async (user: User): Promise<User> => {
  userCache = undefined;
  return user.save();
};

export const generateApiKey = async (name: string): Promise<string> => {
  const user = await getUser(name);
  if (!user) {
    throw new Error("No user found");
  }

  const apiKey = (generateHash(uuid()) + generateHash(uuid())).slice(
    0,
    20 + Math.random() * 100,
  );
  user.apiKey = await hash(apiKey, saltRounds);
  user.apiKeyHash = generateHash(apiKey);
  await updateUser(user);
  return apiKey;
};

export const deleteUser = async (id: number): Promise<User | undefined> => {
  userCache = undefined;
  const user = await User.findOne(id);
  return user?.remove();
};

export const verifyPassword = async (
  name: string,
  secret: string,
): Promise<User | undefined> => {
  const user = await getUser(name);
  if (user) {
    const match = await compare(secret, user.password);
    return match ? user : undefined;
  }
  return undefined;
};

export const verifyApiKey = async (
  secret: string,
): Promise<User | undefined> => {
  const users = await getUsers();
  const [user] = users.filter(({ apiKey, apiKeyHash }) => {
    if (apiKeyHash === generateHash(secret)) {
      return compare(secret, apiKey);
    }
    return false;
  });
  if (user) {
    return user;
  }
  return undefined;
};

export const changePassword = async (
  nameOrId: string | number,
  password: string,
): Promise<void> => {
  const user = await getUser(nameOrId);
  if (!user) {
    throw new Error("No user found");
  }
  user.password = await hash(password, saltRounds);
  await updateUser(user);
};
