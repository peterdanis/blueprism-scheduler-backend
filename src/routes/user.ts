import { addUser, getUser, getUsers } from "../controllers/user";
import { Router } from "express";
import User from "../entities/User";

const router = Router();

router.get("/", async (req, res, next) => {
  try {
    const users = await getUsers();
    res.status(200).json(users);
  } catch (error) {
    next(error);
  }
});

router.post("/", async (req, res, next) => {
  try {
    const { name, password } = req.params;
    if (!name || !password) {
      res.status(422).json({
        error: `User can not be created, ${
          name ? "password" : "name"
        } paremeter is missing`,
      });
      return;
    }
    const user = await addUser(name, password);
    res.status(200).json(user);
  } catch (error) {
    next(error);
  }
});

router.get("/:userId", async (req, res, next) => {
  try {
    const { userId } = req.params;
    let user: User | undefined;
    if (userId && !Number.isNaN(userId)) {
      user = await getUser(parseInt(userId, 10));
    }
    if (!user) {
      res.status(404).json({ error: "User not found" });
      return;
    }
    res.status(200).json(user);
  } catch (error) {
    next(error);
  }
});

export default router;
