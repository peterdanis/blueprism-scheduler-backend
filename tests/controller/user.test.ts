import { Connection, createConnection } from "typeorm";
import { addUser } from "../../src/controller/user";
import config from "../testUtils/testConnectionConfig";
import User from "../../src/entity/User";

let connection: Connection;
const testPassword = "password#123";

beforeAll(async () => {
  connection = await createConnection(config);
});

afterAll(async () => {
  await connection.close();
});

describe("User", () => {
  test("add new user", async () => {
    await addUser("testUser", testPassword);
    const users = await User.find();
    expect(users).toMatchSnapshot();
  });
});
