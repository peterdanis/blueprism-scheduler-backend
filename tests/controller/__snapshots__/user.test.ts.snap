// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User add new user 1`] = `
Array [
  User {
    "admin": false,
    "id": 1,
    "name": "testUser",
    "password": "$2b$10$wNSddHlUh3CBnrIv89UwruhKYY/fbMXDn944DvSMurQCleby/9Bqe",
  },
]
`;
